crates/core/src/client.rs

```diff
@@ -0,0 +1,252 @@
+                let local_block_num = node.read().await.l1_block_num;
+
+                if local_block_num < sn_block_num {
+                    // TODO: Fetch and hash the headers for the amout of blocks l1 is behind l2
+                    match l2_client.get_block_with_tx_hashes(BlockId::Tag(SnBlockTag::Latest)).await.unwrap() {
+                        MaybePendingBlockWithTxHashes::Block(block) => {
+                            info!(
+                                "{} blocks behind - L1 block #({sn_block_num}) L2 block #({:?})",
+                                block.block_number - sn_block_num,
+                                block.block_number
+                            );
+
+                            let mut node_lock = node.write().await;
+                            node_lock.l1_state_root = sn_root;
+                            node_lock.l1_block_num = sn_block_num;
+                        }
+                        MaybePendingBlockWithTxHashes::PendingBlock(_) => warn!("expecting latest got pending"),
+                    };
+                }
+
+                thread::sleep(time::Duration::from_secs(config.poll_secs));
+            }
+        };
+        #[cfg(not(target_arch = "wasm32"))]
+        task::spawn(state_loop);
+        #[cfg(target_arch = "wasm32")]
+        wasm_bindgen_futures::spawn_local(state_loop);
+
+        Ok(())
+    }
+
+    pub async fn sn_state_root(&self) -> Result<FieldElement> {
+        sn_state_root_inner(&self.helios_client, self.core_contract_addr).await
+    }
+
+    pub async fn sn_state_block_number(&self) -> Result<u64, CoreError> {
+        sn_state_block_number_inner(&self.helios_client, self.core_contract_addr).await
+    }
+
+    pub async fn sn_state_block_hash(&self) -> Result<FieldElement, CoreError> {
+        let data = StateBlockHashCall::selector();
+        let call_opts = simple_call_opts(self.core_contract_addr, data.into());
+
+        let sn_block_hash =
+            self.helios_client.call(&call_opts, BlockTag::Latest).await.map_err(CoreError::FetchL1Val)?;
+
+        FieldElement::from_byte_slice_be(&sn_block_hash).map_err(|e| CoreError::FetchL1Val(eyre!("{e}")))
+    }
+
+    pub async fn get_local_root(&self) -> FieldElement {
+        self.node.read().await.l1_state_root
+    }
+
+    pub async fn get_local_block_num(&self) -> u64 {
+        self.node.read().await.l1_block_num
+    }
+
+    pub async fn get_local_block_id(&self, provided_block_id: BlockId) -> BlockId {
+        let local_block_num = self.get_local_block_num().await;
+
+        // allow for historical block ids
+        match provided_block_id {
+            BlockId::Number(num) => {
+                if local_block_num > num {
+                    provided_block_id
+                } else {
+                    BlockId::Number(local_block_num)
+                }
+            }
+            _ => BlockId::Number(local_block_num),
```

Comment:
Can arg `BlockId` be hash for older block then local?

Reply:
It can. I think we need to implement the local cache of block data first as a first big ticket after this merge then add that as an API on top of it

Reply:
To be handled in:

https://github.com/keep-starknet-strange/beerus/issues/550